mailboxes/folders-lib.pl

```diff
@@ -2798,17 +2798,31 @@ sub disable_html_images
 	my ($before, $allimg, $img, $after) = ($1, $2, $3, $4);
 	$img =~ s/^'(.*)'$/$1/ || $img =~ s/^"(.*)"$/$1/;
 	push(@$urls, $img) if ($urls);
-	if ($dis == 0) {
+	if ($dis == 3) {
+		# Let server load it in async mode
+		my $imgid = substitute_pattern('[a-f0-9]{40}');
+		my $imgcont = $allimg;
+		$imgcont =~ s/src=/data-presrc=/g;
+		$imgcont =~ s/\Q$img\E/$imgid/g;
+		$newhtml .= $before.$imgcont;
+		$masked_img++;
+		# Try switching to remote user to correctly
+		# set temporary directory for super-users
+		&switch_to_remote_user();
+		# Save image URL reference
+		&write_file_contents(&tempname($imgid), $img);
```

Comment:
What's going to clean up these temp files?

Reply:
> What's going to clean up these temp files?

Check this code https://github.com/webmin/webmin/pull/1917/files#diff-f6a5dfbecf6813723404d2314ff7df493f8fe396bcc853566d6d2571795392d7R26-R29.

Reply:
Can you explain to me what this code is doing and how it's used?

Reply:
> Can you explain to me what this code is doing and how it's used?

Yes, with pleasure. I will explain the whole process.

As we read an email and later filter images using `disable_html_images`, and if the mode is set to the new default `view_images=3` then:

  1. If email contains images we are not only extract URLs as before but also write a random temporary file with the given pattern (_hex_ string) in user's home, e.g. `/home/user1/.tmp/webmin/eacadbaaebfe2aafceb8bfff4cb0fcbdcefb316d`
  2. This temporary file has an actual image URL
  3. In place `src=` attribute name of an `<img>`, is replaced with `data-presrc` and attribute value (link like `https://host/image.png`) is replaced with random _hex_ string we generated earlier
  4. We keep doing it further for each found image and we send it as part of HTML that is passed to the `iframe`
  5. Later, when the page with an email is loaded the [JavaScript code](https://github.com/webmin/webmin/pull/1917/files#diff-f89a4d333717390003e161a6a16387b29c9dd6b60f16160cd6ccf7da0c4ec493R2928-R2945) kicks in (all that is described happens within a split second)
  6. This code finds all images in an iframe with `data-presrc` attribute and makes a call to a new XHR API with the given, previously stored _hex_ string to find the temporary file that contains image URL
  7. When the file is found and read we delete it
  8. As we now have image pure URL we jump into downloading the linked file and storing its content in a temporary (trans) file
  9. As we have the image content written to the temp (trans) file we check on its MIME type to be added to the `Content-type:` header
  10. Finally we stream file's content back to the client
  11. Already mentioned JavaScript code receives blob data from the server and uses  `URL.createObjectURL`  to create a link.

That's all.

Reply:
Ok I see ... but instead of using a temp file referenced by a random hex ID, why not just pass the URL to the server and have xhr.cgi download that URL and stream the image back? That would avoid the need to create a temp file..

Reply:
You're right! Funny, I'm not sure why I used temp files at the first place. I will fix it as soon as I get back online.

Reply:
Jamie, check the latest commit. It should fix it and use no temp file.